lgc/builder/llpcYCbCrAddressHandler.cpp

```diff
@@ -0,0 +1,190 @@
+/**
+***********************************************************************************************************************
+* @file  llpcYCbCrAddressHandler.cpp
+* @brief LLPC source file: Implementation of LLPC class YCbCrAddressHandler
+***********************************************************************************************************************
+*/
+#include "llpcYCbCrAddressHandler.h"
+#include "llpcGfxRegHandler.h"
+#include "llpcInternal.h"
+
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+
+using namespace lgc;
+using namespace llvm;
+
+// =====================================================================================================================
+// Generate base address of each image plane
+void YCbCrAddressHandler::GenBaseAddress(
+    uint32_t planeNum) // The count of planes
+{
+    if (planeNum > 0)
+    {
+        // m_PlaneBaseAddresses[0] is the same as original base addressed as passed in SRD
+        m_PlaneBaseAddresses.push_back(m_pRegHelper->GetReg(SqRsrcRegs::BaseAddress));
+        if (planeNum > 1)
+        {
+            // m_PlaneBaseAddresses[1] = m_PlaneBaseAddresses[0] + addr_256B(m_pPitchY * m_pHeightY)
+            m_PlaneBaseAddresses.push_back(m_pBuilder->CreateAdd(m_PlaneBaseAddresses[0],
+                                           m_pBuilder->CreateLShr(m_pBuilder->CreateMul(m_pPitchY, m_pHeightY),
+                                           m_pBuilder->getInt32(8))));
+            if (planeNum > 2)
+            {
+                // m_PlaneBaseAddresses[2] = m_PlaneBaseAddresses[1] + addr_256B(m_pPitchCb * m_pHeightCb)
+                m_PlaneBaseAddresses.push_back(m_pBuilder->CreateAdd(m_PlaneBaseAddresses[1],
+                                               m_pBuilder->CreateLShr(m_pBuilder->CreateMul(m_pPitchCb, m_pHeightCb),
+                                               m_pBuilder->getInt32(8))));
+            }
+        }
+    }
+}
+
+// =====================================================================================================================
+// Power2align operation
+Value* YCbCrAddressHandler::Power2Align(
+    Value*   pX,    // [in] Value needs to be aligned
+    uint32_t align) // Align base
+{
+    uint32_t alignM1 = align - 1;
+    uint32_t alignM1Inv = ~alignM1;
+    Value* pResult = m_pBuilder->CreateAdd(pX, m_pBuilder->getInt32(alignM1));
+    return m_pBuilder->CreateAnd(pResult, m_pBuilder->getInt32(alignM1Inv));
+}
+
+// =====================================================================================================================
+// Calculate height and pitch
+// Note: Need pRegHelper to be bounded
+void YCbCrAddressHandler::GenHeightAndPitch(
+    uint32_t bits,          // Channel bits
+    uint32_t bpp,           // Bits per pixel
+    uint32_t xBitCount,     // Effective channel bits
+    bool     isTileOptimal, // Is tiling optimal
+    uint32_t planeNum)      // Plane counts
```

Comment:
nit: maybe `// Number of planes`, to keep it consistent with the argument name?
Are all uint32 values valid here?

Reply:
Modified.